beacon-chain.k

```diff
@@ -1398,12 +1406,16 @@ def get_attestation_deltas(state: BeaconState) -> Tuple[Sequence[Gwei], Sequence
                        UnslashedIndices,
                        AttestingBalance,
                        #mapMapPair(Rewards => #if contains(I, UnslashedIndices)
-                                              #then Rewards[I <- ({Rewards[I]}:>Int +Int get_base_reward(I)
-                                                                  *Int AttestingBalance /Int TotalBalance)]
+                                              #then Rewards[I <- ei(0, 2 ^Int 64 -Int 1, ci({Rewards[I]}:>Int) +UInt64 get_base_reward(I)
```

Comment:
I think `ci({Rewards[I]}:>Int)` should be changed to `ui(MIN, MAX, {Rewards[I]}:>Int)`, with proper MIN/MAX bounds.  The thing is that `ci(_)` is meant to be used only for constants, but `Rewards[I]` is not a constant.  I guess the bounds can be given the same bounds of `get_base_reward(I)`?

Reply:
One way to give the same bounds of `get_base_reward(...)` to `Rewards[I]` is that you come up with a reasonable MIN/MAX bounds, and ensure that `get_base_reward(...)` indeed falls within the bounds, like:
```
rule get_base_reward(...) => ensureBounds(MIN, MAX, ...)
```
where `ensureBounds` is similar to `ei` but preserves the UInt64 value.
```
syntax UInt64 ::= ensureBounds(Int, Int, UInt64) [function]
rule ensureBounds(L, U, #uint64(L0, U0, V)) => #uint64(L0, U0, V) requires L <=Int L0 andBool U0 <=Int U
```

Reply:
I agree that `Rewards[I]` should be given bounds. But I don't think the appropriate bound would be the same as get_base_reward.
There seems to be an error in the k specs here, which was partly present in the [python specs](https://github.com/ethereum/eth2.0-specs/pull/1682/files) as well, in that we should be adding to the rewards for every attestation (right now we are resetting it).

With the correct implementation, it seems that the appropriate max bound for Rewards would need to be:
```
get_base_reward(state, index) * (attesting_balance // total_balance) * (ATTESTATIONS - 1)
```

where ATTESTATIONS would be the upper bound of `length(matching_source_attestations) + length(matching_target_attestations) + length(matching_head_attestations)` which is bounded to `3 * MAX_ATTESTATIONS * SLOTS_PER_EPOCH`